### PR TITLE
Add Transaction Reporting functionality

### PR DESCRIPTION
**Pull request by PatrickCronin**
In order to encourage more usage of the [minFraud Report Transaction API](https://dev.maxmind.com/minfraud/report-transaction/), this PR adds a method to this client library that simplifies transaction reporting.

### CODE REVIEW

File: tests/MaxMind/Test/ServiceClientTest.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        $curlVersion = curl_version();
+        $factory->expects($this->exactly($callsToRequest))
+            ->method('request')
+            ->with(
+                $this->equalTo($url),
+                $this->equalTo(
+                    [
+                        'headers' => $headers,
+                        'userAgent' => 'minFraud-API/' . $class::VERSION
```
Comment (PatrickCronin) [inline tests/MaxMind/Test/ServiceClientTest.php:65]: I want to point out `$class::VERSION` here -- there's no corresponding `use` statement at the top, so I'm not sure how it works. Tests do cover the user agent field though, so it's working, but I'm not sure how.

Comment (oschwald) [reply tests/MaxMind/Test/ServiceClientTest.php:65]: You are passing the fully qualified name.

File: src/MinFraud.php [hunk 2]
```diff
@@ -36,17 +34,10 @@
  *
  * If the request fails, an exception is thrown.
  */
-class MinFraud
+class MinFraud extends ServiceClient
 {
-    const VERSION = 'v1.12.0';
```
Comment (oschwald) [inline src/MinFraud.php:41]: We should update `dev-bin/release.sh` for this change (and presumably that `sed` fix as well).

File: src/MinFraud/Validation/Rules/TransactionReport.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public function __construct()
+    {
+        $this->validatable = v::keySet(
+            v::key('chargeback_code', v::stringType(), false),
+            v::key('ip_address', v::ip(), true),
+            v::key('maxmind_id', v::stringType()->length(8, 8), false),
+            v::key(
+                'minfraud_id',
+                v::regex('/^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$/'),
```
Comment (oschwald) [inline src/MinFraud/Validation/Rules/TransactionReport.php:21]: You could simplify this by using the `i` option.

File: src/ReportTransaction.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+     *                                 data
+     * @throws AuthenticationException when there is an issue authenticating the
+     *                                 request
+     * @throws InvalidRequestException when the request is invalid for some
+     *                                 other reason, e.g., invalid JSON in the POST.
+     * @throws HttpException           when an unexpected HTTP error occurs
+     * @throws WebServiceException     when some other error occurs. This also
+     *                                 serves as the base class for the above exceptions.
+     */
+    public function reportTransaction($values)
```
Comment (oschwald) [inline src/ReportTransaction.php:48]: This stutters a bit `ReportTransaction->new(..)->reportTransaction`. That said, I see why you picked it given the minFraud ones are named after the endpoint. Maybe `report` would be sufficient? Not sure.

File: src/ReportTransaction.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        $values = $this->cleanAndValidate('TransactionReport', $values);
+
+        if (!isset($values['ip_address'])) {
+            throw new InvalidInputException('Key ip_address must be present in request');
+        }
+        if (!isset($values['tag'])) {
+            throw new InvalidInputException('Key tag must be present in request');
+        }
```
Comment (oschwald) [inline src/ReportTransaction.php:57]: Not sure these are entirely necessary given that the validator has them marked as required. However, I see that we do it for `/device/ip_address`. I don't really recall why. Maybe we wanted an extra check even when validation is disabled or maybe the requirement stuff wasn't working in Respect at the time.

File: tests/MaxMind/Test/ReportTransactionTest.php [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+                Data::minimalRequest(),
+                1
+            )->reportTransaction($req),
+            'response from request including nulls'
+        );
+    }
+
+    /**
+     * @expectedException \MaxMind\Exception\InvalidInputException
+     * @expectedExceptionMessage Must have keys
```
Comment (oschwald) [inline tests/MaxMind/Test/ReportTransactionTest.php:57]: I think these [annotations are deprecated](https://github.com/sebastianbergmann/phpunit/issues/3332) in new PHPUnit versions. We should be moving to the method form of testing. See [the docs](https://phpunit.readthedocs.io/en/9.0/writing-tests-for-phpunit.html#testing-exceptions).

File: tests/MaxMind/Test/ReportTransactionTest.php [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        $this->createReportTransactionRequest(
+            $req,
+            0
+        )->reportTransaction($req);
+    }
+
+    public function notStringTypes()
+    {
+        return [
+            [1],
```
Comment (oschwald) [inline tests/MaxMind/Test/ReportTransactionTest.php:182]: If we've bothered to create a data provider, we might as well add some other data types. Maybe an array or an object?

File: tests/MaxMind/Test/ServiceClientTest.php (hunk 1 above)
Comment (PatrickCronin) [reply tests/MaxMind/Test/ServiceClientTest.php:65]: Ah, OK, thanks!

File: src/ReportTransaction.php (hunk 4 above)
Comment (PatrickCronin) [reply src/ReportTransaction.php:48]: How about `sendReport`?

File: src/ReportTransaction.php (hunk 5 above)
Comment (PatrickCronin) [reply src/ReportTransaction.php:57]: My thoughts are along the lines of your first thought -- that regardless of whether validation is enabled or not, we'll avoid the obvious issues with missing required fields. I'm leaning towards leaving them to maintain consistency in the library's behavior. If you want them out, please say so.

File: src/ReportTransaction.php (hunk 4 above)
Comment (oschwald) [reply src/ReportTransaction.php:48]: If we are going to break consistency, which I think is fine here, how about just `send`?

File: tests/MaxMind/Test/ReportTransactionTest.php (hunk 6 above)
Comment (PatrickCronin) [reply tests/MaxMind/Test/ReportTransactionTest.php:57]: I've updated to the method form for `expectException`, `expectExceptionMessage` and `expectExceptionMessageRegExp`. Note that the last one [is deprecated now too](https://github.com/sebastianbergmann/phpunit/commit/d1199cb2e43a934b51521656be9748f63febe18e), and I tried to use it's successor `expectExceptionMessageMatches`, but it doesn't exist in the version of phpunit we've got listed in composer.json.

File: src/ReportTransaction.php (hunk 4 above)
Comment (PatrickCronin) [reply src/ReportTransaction.php:48]: I've gone back to `report` as that's the operative verb in the API's name. I hope that's OK.

File: README.md [hunk 8]
```diff
@@ -77,7 +77,13 @@ Page](https://maxmind.github.io/minfraud-api-php/) under the "API" tab.
 
 ## Usage ##
 
-To use this API, create a new `\MaxMind\MinFraud` object. The constructor
+This library provides access to both the [minFraud2 (Score, Insights and
```
Comment (oschwald) [inline README.md:80]: "minFraud2" is an internal name. We should just use "minFraud" in public documentation. The same applies below.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+1.13.0 (2020-05-14)
+
```
Comment (oschwald) [inline CHANGELOG.md:5]: You are missing the `-`. Also, usually we leave out the release date until we are ready to actually release, unless it is a given that we will release on that day.

File: README.md [hunk 10]
```diff
@@ -243,6 +249,72 @@ foreach ($scoreResponse->warnings as $warning) {
 }
 ```
 
+### Report Transactions API ###
+
+MaxMind encourages the use of this API as data received through this channel is
+used to continually improve the accuracy of our fraud detection algorithms.
+
+To use the Report Transactions API, create a new `\MaxMind\ReportTransaction`
```
Comment (oschwald) [inline README.md:257]: I didn't notice this before, but I wonder if `\MaxMind\MinFraud\ReportTransaction` would be a better name. It is a minFraud API, just not one of our scoring APIs.

Similarly, `\MaxMind\ServiceClients` could lead to accidental name collisions with our other client APIs.

File: README.md [hunk 11]
```diff
@@ -243,6 +249,72 @@ foreach ($scoreResponse->warnings as $warning) {
...
+require_once 'vendor/autoload.php';
+use MaxMind\ReportTransaction;
+
+# The constructor for ReportTransaction takes your account ID, your license key,
+# and optionally an array of options.
+$rt = new ReportTransaction(1, 'ABCD567890');
+
+$rt->report(
+	'ip_address'      => '81.2.69.160',
+    'tag'             => 'chargeback',
```
Comment (oschwald) [inline README.md:308]: The formatting on this looks off in GitHub. Maybe you are switching between tabs and spaces?

File: tests/MaxMind/Test/MinFraudTest.php [hunk 12]
```diff
@@ -145,30 +142,32 @@ private function createNullRequest()
...
-     * @expectedExceptionMessage Must have keys
      * @dataProvider services
      *
      * @param mixed $class
      * @param mixed $service
      */
     public function testMissingIpAddress($class, $service)
     {
+        $this->expectException(InvalidInputException::class);
+        $this->expectExceptionMessage('Must have keys');
```
Comment (oschwald) [inline tests/MaxMind/Test/MinFraudTest.php:153]: Nice. I wasn't saying you had to do these ones as part of the story, but it is good that we won't have to deal with them later.

File: README.md (hunk 10 above)
Comment (PatrickCronin) [reply README.md:257]: I've made this change.

File: tests/MaxMind/Test/MinFraudTest.php (hunk 12 above)
Comment (PatrickCronin) [reply tests/MaxMind/Test/MinFraudTest.php:153]: Yeah, no worries. It was just a couple regexp finds/replaces.

File: dev-bin/release.sh [hunk 13]
```diff
@@ -44,7 +44,7 @@ fi
 php composer.phar self-update
 php composer.phar update --no-dev
 
-perl -pi -e "s/(?<=const VERSION = ').+?(?=';)/$tag/g" src/MinFraud.php
+perl -pi -e "s/(?<=const VERSION = ').+?(?=';)/$tag/g" src/ServiceClient.php
```
Comment (oschwald) [inline dev-bin/release.sh:47]: This needs to be updated now.

Comment (PatrickCronin) [reply dev-bin/release.sh]: Thanks -- great catch! I've rebased this in with the initial abstraction commit.

**Pull request merged**